### ISSUE TITLE
Recursive flatten

### ISSUE THREAD

**Issue by edsilv** [jmespath/jmespath.js#6]
Hi,

Is it possible to retrieve a flattened list of nodes recursively for a tree of arbitrary depth? I know that the flatten operator isn't recursive, but is there some other pattern that can be used?

So for this data:

    {
      "node": {
         "label": "root node",
         "nodes": [
           {
             "label": "node 1",
             "nodes": [
               {
                 "label": "node 1:1"
               },
               {
                 "label": "node 1:2"
               }
             ]
           },
           {
             "label": "node 2",
             "nodes": [
               {
                 "label": "node 2:1"
               },
               {
                 "label": "node 2:2"
               }
             ]
           }
         ] 
       }
    }

I could get:

    [
      {
        "label": "root node" 
      },
      {
        "label": "node 1" 
      },
      {
        "label": "node 1:1" 
      },
      {
        "label": "node 1:2" 
      },
      {
        "label": "node 2" 
      },
      {
        "label": "node 2:1" 
      },
      {
        "label": "node 2:2" 
      }
    ]

**Comment by jamesls** [comment]
That's not currently possible, but I think an interesting addition.  Perhaps it's worth adding some sort of recursive flatten function?

cc @mtdowling

**Comment by mtdowling** [comment]
I think a recursive flatten is too specific. I do think that a more general way to extract elements recursively would be interesting though.

**Comment by jamesls** [comment]
What about something like `recurse(start, expref_traverse, expref_select)` where:

* start, the root element to start
* expref_traverse - the expref to recurse into.  at each level this expref is applied to the current object and if it evalutes to anything, then that result is subsequently traversed
* expref_select - at each level, this is the value to extract and return.

So a recursive flatten in this scenario would look like `recurse(node, &nodes, &label)`.

**Comment by mtdowling** [comment]
That's kinda cool. Reminds me of clojure's `tree-seq` a bit. https://clojuredocs.org/clojure.core/tree-seq

Note that your output would probably be a list of `["root node", "node 1", "node 2" ...]` and not a list of hashes.

**Comment by highfield** [comment]
Please, consider the Linq-way as well:
http://blogs.interknowlogy.com/2008/10/10/use-linqs-selectmany-method-to-flatten-collections/

**Comment by edsilv** [comment]
Here is a TypeScript implementation of LINQ-style tree traversal:

https://github.com/BSick7/exjs/blob/f607b01d31d775730b9c1fdb168e271cb2c8dbe3/src/traverse.ts

exjs is really powerful, but jmespath has a much more accessible string-based query language. It would be great if there were a jmespath wrapper for exjs.

**Comment by highfield** [comment]
I didn't mean another library, rather the Linq "select-many" approach/syntax to trasverse a tree. Better: to map a dual-level structure.

**Comment by edsilv** [comment]
Yep, I posted it as a reference. From a beginners' point of view, I'd be interested to know how jmespath compares to LINQ - is it essentially the same concepts just expressed through a different query language? Perhaps a question more suited for the gitter.im chat.

**Comment by johnt337** [comment]
Is this something that has any interest?

**Comment by serge-medvedev** [comment]
Yeah, would be great to have.

**Comment by callowaylc** [comment]
guys, just use jq and let this awful tool die.. lol, why on earth would anyone want to flatten a json result set so that it can be run through a shell pipeline.. I mean, that's just crazy talk azure cli.

**Comment by jamesls** [comment]
I can understand being frustrated, but please keep it cordial if you want to comment on these issues.
